src/app/shared/ui/tournament-bracket.component.ts

```diff
@@ -0,0 +1,203 @@
+  private readonly eventService = inject(EventService);
+
+  private readonly eventTeams = computed(() =>
+    this.eventService
+      .eventTeamsWithParticipantInfo()
+      ?.filter((eventTeam) => eventTeam.event_id === this.eventId()),
+  );
+
+  readonly eventId = input.required<number>();
+
+  readonly bracket = computed(() => this.generateBracket(this.eventTeams()));
+  readonly selectedNodes: TreeNode<Partial<EventTeamModelWithWinnerFlag>>[] =
+    [];
+
+  generateBracket(
+    eventTeams: EventTeamModel[] | undefined,
+  ): TreeNode<Partial<EventTeamModelWithWinnerFlag>>[] {
+    if (!eventTeams) {
+      return [{}];
+    }
+
+    const nodeTemplate = { expanded: true, type: 'node' };
+    const bracket = this.generateBracketRecursively(
+      { data: { seed: 1, isWinner: false } },
+      eventTeams.length,
+      1,
+      eventTeams,
+      nodeTemplate,
+      false,
+    );
+    return [bracket];
+  }
+
+  private generateBracketRecursively(
+    parentNode: TreeNode<Partial<EventTeamModelWithWinnerFlag>>,
+    numberOfTeams: number,
+    numberOfTeamsInRound: number,
+    eventTeams: EventTeamModel[],
+    nodeTemplate: TreeNode,
+    isLowSeed: boolean,
+  ): TreeNode<Partial<EventTeamModelWithWinnerFlag>> {
+    const bracketNode: TreeNode<Partial<EventTeamModelWithWinnerFlag>> = {
+      ...nodeTemplate,
+      data: {
+        seed: isLowSeed
+          ? parentNode.data!.seed
+          : numberOfTeamsInRound + 1 - parentNode.data!.seed!,
+        isWinner: false,
+      },
+    };
+
+    if (
+      2 * numberOfTeamsInRound + 1 - bracketNode.data!.seed! >
+      numberOfTeams
+    ) {
+      bracketNode.data = {
+        ...eventTeams.find(
+          (eventTeam) => eventTeam.seed === bracketNode.data!.seed,
+        ),
+      };
+      return bracketNode;
+    }
+
+    bracketNode.children = [
+      this.generateBracketRecursively(
+        bracketNode,
+        numberOfTeams,
+        2 * numberOfTeamsInRound,
+        eventTeams,
+        nodeTemplate,
+        false,
+      ),
+      this.generateBracketRecursively(
+        bracketNode,
+        numberOfTeams,
+        2 * numberOfTeamsInRound,
+        eventTeams,
+        nodeTemplate,
+        true,
+      ),
+    ];
+
+    delete bracketNode.data!.seed;
+
+    return bracketNode;
+  }
+
+  setMatchWinner(
+    ev: TreeNodeSelectEvent,
+    bracket: TreeNode<Partial<EventTeamModelWithWinnerFlag>>[],
+    selectedNodes: TreeNode<Partial<EventTeamModelWithWinnerFlag>>[],
+  ) {
+    const parent = this.getNodeParent(ev.node, bracket[0]);
+
+    if (parent) {
+      parent.data = ev.node.data;
+    }
+
+    const sibling = parent?.children?.find((child) => child !== ev.node);
+
+    const siblingIndex =
+      selectedNodes.findIndex((node) => node === sibling) ?? -1;
+
+    if (siblingIndex !== -1) {
+      selectedNodes.splice(siblingIndex, 1);
+    }
+  }
+
+  private getNodeParent(
+    node: TreeNode,
+    rootNode: TreeNode,
+  ): TreeNode | undefined {
```

Comment:
This is fine, but it's so silly that we need to do this. `parent` always being `undefined` is probably a bug in PrimeNG. Maybe we could file a bug report at some point.

Reply:
Agreed, I tried a few different ways of assigning `parent` with no luck.  I can't say I'm 100% certain I wasn't doing something wrong, but I'm pretty sure this is a bug.